doc/bookmarkctl.1
.TH BOOKMARKCTL 1 "Dec 01, 2024" 0.1.0 "BookmarkFS User Manual"
.
.SH NAME
bookmarkctl - manage a mounted BookmarkFS filesystem
.
.SH SYNOPSIS
.B bookmarkctl
.I subcmd
.RI [ args ]
.PP
.B bookmarkctl
.B permd
.RI [ options ]
.I name1
.I name2
.I pathname
.PP
.B bookmarkctl
.B fsck
.I pathname
.PP
.B bookmarkctl
.B xattr\-list
.I pathname
.PP
.B bookmarkctl
.B xattr\-get
.RI [ options ]
.I attrname
.IR pathname ...
.PP
.B bookmarkctl
.B xattr\-get
.RI [ options ]
.B \-a
.IR attrname "... " pathname
.PP
.B bookmarkctl
.B xattr\-set
.RI [ options ]
.I attrname
.I pathname
.PP
.B bookmarkctl
.B help
.PP
.B bookmarkctl
.B version
.
.SH DESCRIPTION
The
.B bookmarkctl
program is a command-line wrapper for various operations on a BookmarkFS
filesystem that cannot be done portably with POSIX utilities alone,
including I/O controls and managing extended attributes.
.SS Permute directory entries
The
.B permd
sub-command re-arranges the order of the directory entries to be returned
from future
.BR readdir (3)
calls.
.PP
The
.I pathname
argument is the path to the directory.
The
.I name1
and
.I name2
arguments are filenames under that directory.
.SS Filesystem check
The
.B fsck
sub-command displays a list of filesystem errors found under the directory
referred to by
.IR pathname .
It does not recurse into subdirectories.
.PP
For the full fsck functionalities, use
.BR fsck.bookmarkfs (1).
.SS List extended attribute names
The
.B xattr\-list
sub-command displays a list of extended attribute names
of the file referred to by
.IR pathname .
.PP
Only the names prefixed with "user.bookmarkfs." are listed,
with the prefix removed.
.SS Get extended attribute value
The
.B xattr\-get
sub-command displays one or multiple extended attribute values
of the file referred to by
.IR pathname .
.PP
The
.I attrname
argument is the name of the extended attribute,
without the "user.bookmarkfs." prefix.
.SS Set extended attribute value
The
.B xattr\-set
sub-command updates the value of an extended attribute
of the file referred to by
.IR pathname .
.PP
The
.I attrname
argument is the name of the extended attribute,
without the "user.bookmarkfs." prefix.
.
.SH OPTIONS
.SS Permute directory entries
.TP
.B \-s
Exchange the positions of the directory entries represented by
.I name1
and
.IR name2 .
.TP
.B \-b
Move the directory entry represented by
.I name1
to the position just before the one represented by
.IR name2 .
.TP
.B \-a
Move the directory entry represented by
.I name1
to the position just after the one represented by
.IR name2 .
.PP
The
.BR \-s ", " \-b
and
.B \-a
options are mutually exclusive.
If multiple options are provided, the last one takes effect.
.SS Get extended attribute value
.TP
.B \-b
Treat the value as binary, and print it verbatim.
.IP
If this option is not provided, non-printable characters are replaced with '?'.
.TP
.B \-a
Switch to multi-attrname mode, where multiple extended attribute names
can be specified instead of multiple files.
.IP
Also, the corresponding
.I attrname
is printed alongside with the value, instead of
.IR pathname .
.TP
.B \-q
Print values only, do not print names.
.IP
If provided twice, do not print the suffix character after each value.
.TP
\fB\-s\fR \fIchar\fR
The character separating each name and value.
Defaults to the ASCII HT character.
.TP
\fB\-n\fR \fIchar\fR
The character suffixing each value.
Defaults to the ASCII LF character.
.PP
For the
.B \-s
and
.B \-n
options, if
.I char
contains multiple characters, the first one is used.
If empty, the ASCII NUL character is used.
.SS Set extended attribute value
.TP
\fB\-v\fR \fIvalue\fR
New value for the extended attribute.
.IP
If this option is not provided, the new value is read from standard input.
.
.SH EXIT STATUS
.TP
0
The operation completed successfully.
.TP
1
An error occurred.
.
.SH NOTES
.SS Availability on FreeBSD
Before 16.0 (not yet released), the FreeBSD
.BR fusefs (5)
implementation does not support
.BR FUSE_IOCTL .
.PP
On such systems, the
.B permd
and
.B fsck
sub-commands (which depends on
.BR ioctl (2))
does not work.
.
.SH SEE ALSO
.BR fsck.bookmarkfs (1),
.BR mount.bookmarkfs (1),
.BR ioctl (2),
.BR readdir (3),
.BR fusefs (5),
.BR xattr (7)
.PP
The full BookmarkFS User Manual is maintained as a Texinfo document.
To read it locally, run:
.PP
.in +4n
.EX
.B info bookmarkfs
.EE
.in
.
.SH COPYRIGHT
Copyright (C) 2024  CismonX <admin@cismon.net>
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.3
or any later version published by the Free Software Foundation;
with no Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.
.PP
You should have received a copy of the license along with this document.
If not, see
.UR https://www.gnu.org/licenses/fdl-1.3.html
.UE .


src/bookmarkctl.c
/**
 * bookmarkfs/src/bookmarkctl.c
 * ----
 *
 * Copyright (C) 2024  CismonX <admin@cismon.net>
 *
 * This file is part of BookmarkFS.
 *
 * BookmarkFS is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BookmarkFS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with BookmarkFS.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <fcntl.h>
#include <sys/uio.h>
#include <unistd.h>

#include "frontend_util.h"
#include "fsck_util.h"
#include "ioctl.h"
#include "macros.h"
#include "version.h"
#include "xattr.h"
#include "xstd.h"

struct xattr_get_ctx {
    char const *prefix;
    unsigned    quiet;
    char        sep;
    char        eol;

    unsigned binary     : 1;
    unsigned multi_attr : 1;
};

// Forward declaration start
static int  dispatch_subcmds  (int, char *[]);
static void print_help        (void);
static void print_version     (void);
static int  subcmd_fsck       (int, char *[]);
static int  subcmd_permd      (int, char *[]);
static int  subcmd_xattr_get  (int, char *[]);
static int  subcmd_xattr_list (int, char *[]);
static int  subcmd_xattr_set  (int, char *[]);
static int  xattr_get_cb      (void *, void *, size_t);
static int  xattr_get_one     (char const *, char *[], int,
                               struct xattr_get_ctx *);
static int  xattr_list_cb     (void *, void *, size_t);
// Forward declaration end

static int
dispatch_subcmds (
    int   argc,
    char *argv[]
) {
    if (--argc < 1) {
        log_puts("command not given; run 'bookmarkctl help' for usage");
        return -1;
    }
    char const *cmd = *(++argv);

    int status = 0;
    if (0 == strcmp("permd", cmd)) {
        status = subcmd_permd(argc, argv);
    } else if (0 == strcmp("xattr-get", cmd)) {
        status = subcmd_xattr_get(argc, argv);
    } else if (0 == strcmp("xattr-set", cmd)) {
        status = subcmd_xattr_set(argc, argv);
    } else if (0 == strcmp("xattr-list", cmd)) {
        status = subcmd_xattr_list(argc, argv);
    } else if (0 == strcmp("fsck", cmd)) {
        status = subcmd_fsck(argc, argv);
    } else if (0 == strcmp("help", cmd)) {
        print_help();
    } else if (0 == strcmp("version", cmd)) {
        print_version();
    } else {
        log_printf("bad command '%s'; run 'bookmarkctl help' for usage", cmd);
        status = -1;
    }
    return status;
}

static void
print_help (void)
{
    puts("Usage: bookmarkctl <cmd> [args]\n"
            "\n"
            "Main commands:\n"
            "  permd         Permute directory entries\n"
            "  fsck          Check filesystem\n"
            "  xattr-list    List extended attribute names\n"
            "  xattr-get     Get extended attribute value\n"
            "  xattr-set     Set extended attribute value\n"
            "\n"
            "Other commands:\n"
            "  help       Print help message\n"
            "  version    Print version information\n"
            "\n"
            "See the bookmarkctl(1) manpage for more information,\n"
            "or run 'info bookmarkfs' for the full user manual.\n"
            "\n"
            "Project homepage: <" BOOKMARKFS_HOMEPAGE_URL ">.");
}

static void
print_version (void)
{
    printf("bookmarkctl (BookmarkFS) %d.%d.%d\n",
            BOOKMARKFS_VER_MAJOR, BOOKMARKFS_VER_MINOR, BOOKMARKFS_VER_PATCH);
}

static int
subcmd_fsck (
    int   argc,
    char *argv[]
) {
    if (--argc != 1) {
        if (argc == 0) {
            log_puts("fsck: pathname must be specified");
        } else {
            log_puts("fsck: too many arguments");
        }
        return -1;
    }
    char const *path = *(++argv);

    int dirfd = open(path, O_RDONLY | O_DIRECTORY);
    if (dirfd < 0) {
        log_printf("open(): %s: %s", path, strerror(errno));
        return -1;
    }

    int status = -1;
    do {
        struct bookmarkfs_fsck_data fsck_data;
        status = ioctl(dirfd, BOOKMARKFS_IOC_FSCK_NEXT, &fsck_data);
        if (status < 0) {
            log_printf("ioctl(): %s", strerror(errno));
            break;
        }
        if (0 != print_fsck_result(status, &fsck_data)) {
            break;
        }
    } while (status != BOOKMARKFS_FSCK_RESULT_END);

    close(dirfd);
    return status;

}

static int
subcmd_permd (
    int   argc,
    char *argv[]
) {
    int op = -1;

    OPT_START(argc, argv, "sba")
    OPT_OPT('s') {
        op = BOOKMARKFS_PERMD_OP_SWAP;
        break;
    }
    OPT_OPT('b') {
        op = BOOKMARKFS_PERMD_OP_MOVE_BEFORE;
        break;
    }
    OPT_OPT('a') {
        op = BOOKMARKFS_PERMD_OP_MOVE_AFTER;
        break;
    }
    OPT_END

    if (op < 0) {
        log_puts("permd: no operation specified");
        return -1;
    }
    struct bookmarkfs_permd_data permd_data;
    permd_data.op = op;

    if (argc != 3) {
        if (argc < 3) {
            log_puts("permd: dentry names and dir path must be specified");
        } else {
            log_puts("permd: too many arguments");
        }
        return -1;
    }
    char const *name1 = argv[0];
    char const *name2 = argv[1];
    char const *path  = argv[2];

#define COPY_NAME(dst, src)                                        \
    if ((dst) + sizeof(dst) == stpncpy(dst, src, sizeof(dst))) {   \
        log_printf("permd: %s: %s", src, strerror(ENAMETOOLONG));  \
        return -1;                                                 \
    }
    COPY_NAME(permd_data.name1, name1);
    COPY_NAME(permd_data.name2, name2);

    int dirfd = open(path, O_RDONLY | O_DIRECTORY);
    if (dirfd < 0) {
        log_printf("open(): %s: %s", path, strerror(errno));
        return -1;
    }

    int status = ioctl(dirfd, BOOKMARKFS_IOC_PERMD, &permd_data);
    if (status < 0) {
        log_printf("ioctl(): %s", strerror(errno));
    }
    close(dirfd);
    return status;
}

static int
subcmd_xattr_get (
    int   argc,
    char *argv[]
) {
    struct xattr_get_ctx ctx = {
        .sep = '\t',
        .eol = '\n',
    };

    OPT_START(argc, argv, "n:bmqs:")
    OPT_OPT('n') {
        ctx.eol = optarg[0];
        break;
    }
    OPT_OPT('b') {
        ctx.binary = 1;
        break;
    }
    OPT_OPT('a') {
        ctx.multi_attr = 1;
        break;
    }
    OPT_OPT('q') {
        ++ctx.quiet;
        break;
    }
    OPT_OPT('s') {
        ctx.sep = optarg[0];
        break;
    }
    OPT_NOVAL
    OPT_END

    if (argc < 2) {
        log_puts("xattr-get: xattr name and file path must be specified");
        return -1;
    }

    if (ctx.multi_attr) {
        return xattr_get_one(argv[argc - 1], argv, argc - 1, &ctx);
    }
    for (int i = 1; i < argc; ++i) {
        char const *path = argv[i];
        if (0 != xattr_get_one(path, argv, 1, &ctx)) {
            return -1;
        }
    }
    return 0;
}

static int
subcmd_xattr_list (
    int   argc,
    char *argv[]
) {
    if (--argc != 1) {
        if (argc < 1) {
            log_puts("xattr-set: file path must be provided");
        } else {
            log_puts("xattr-set: too many arguments");
        }
        return -1;
    }
    char const *path = *(++argv);

    int fd = bookmarkfs_xattr_open(path);
    if (fd < 0) {
        return -1;
    }
    int status = bookmarkfs_xattr_list(fd, xattr_list_cb, NULL);
    close(fd);
    return status;
}

static int
subcmd_xattr_set (
    int   argc,
    char *argv[]
) {
    char *val = NULL;

    OPT_START(argc, argv, "v:")
    OPT_OPT('v') {
        val = optarg;
        break;
    }
    OPT_NOVAL
    OPT_END

    if (argc != 2) {
        if (argc < 2) {
            log_puts("xattr-set: xattr name and file path must be specified");
        } else {
            log_puts("xattr-set: too many arguments");
        }
        return -1;
    }
    char const *name = argv[0];
    char const *path = argv[1];

    int fd = bookmarkfs_xattr_open(path);
    if (fd < 0) {
        return -1;
    }
    int status = -1;

    size_t  val_size = 4096;
    char   *buf      = val;
    if (buf == NULL) {
        for (size_t off = 0; ; off = val_size, val_size += val_size >> 1) {
            buf = xrealloc(buf, val_size);
            ssize_t len = xread(STDIN_FILENO, buf + off, val_size - off);
            if (len < 0) {
                goto end;
            }
            if (off + len < val_size) {
                val_size = off + len;
                break;
            }
        }
    } else {
        val_size = strlen(buf);
    }
    status = bookmarkfs_xattr_set(fd, name, buf, val_size);

  end:
    if (val == NULL) {
        free(buf);
    }
    close(fd);
    return status;
}

static int
xattr_get_cb (
    void   *user_data,
    void   *buf,
    size_t  buf_len
) {
    struct xattr_get_ctx *ctx = user_data;

    struct iovec iov[4] = {
        { (char *)ctx->prefix, 0 }, { &ctx->sep, 0 },
        { buf, buf_len },           { &ctx->eol, 0 },
    };
    if (ctx->quiet < 2) {
        if (ctx->quiet < 1) {
            iov[0].iov_len = strlen(iov[0].iov_base);
            iov[1].iov_len = 1;
        }
        iov[3].iov_len = 1;
    }
    if (!ctx->binary) {
        for (unsigned char *s = buf, *end = s; s < end; ++s) {
            if (iscntrl(*s)) {
                *s = '?';
            }
        }
    }
    return xwritev(STDOUT_FILENO, iov, 4);
}

static int
xattr_get_one (
    char const           *path,
    char                 *names[],
    int                   names_cnt,
    struct xattr_get_ctx *ctx
) {
    int fd = bookmarkfs_xattr_open(path);
    if (fd < 0) {
        return -1;
    }

    int status;
    for (int i = 0; i < names_cnt; ++i) {
        char const *name = names[i];

        ctx->prefix = ctx->multi_attr ? name : path;
        status = bookmarkfs_xattr_get(fd, name, xattr_get_cb, ctx);
        if (status < 0) {
            goto end;
        }
    }

  end:
    close(fd);
    return status;
}

static int
xattr_list_cb (
    void   *UNUSED_VAR(user_data),
    void   *buf,
    size_t  buf_len
) {
    if (0 > printf("%.*s\n", (int)buf_len, (char *)buf)) {
        log_printf("printf(): %s", strerror(errno));
        return -1;
    }
    return 0;
}

int
main (
    int   argc,
    char *argv[]
) {
    if (0 != dispatch_subcmds(argc, argv)) {
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
